src/mem/alloc.h

```diff
@@ -794,16 +805,16 @@ namespace snmalloc
 
     template<Boundary location>
     static uintptr_t
-    external_pointer(void* p, uint8_t sizeclass, size_t end_point)
+    external_pointer(void* p, uint8_t sizeclass, void* end_point)
     {
       size_t rsize = sizeclass_to_size(sizeclass);
-      size_t end_point_correction = location == End ?
-        (end_point - 1) :
-        (location == OnePastEnd ? end_point : (end_point - rsize));
-      size_t offset_from_end =
-        (end_point - 1) - static_cast<size_t>(address_cast(p));
-      size_t end_to_end = round_by_sizeclass(rsize, offset_from_end);
-      return end_point_correction - end_to_end;
+      void *end_point_correction = location == End ?
+        ((char*)end_point - 1) :
+        (location == OnePastEnd ? end_point : pointer_offset(end_point,-rsize));
+      ptrdiff_t offset_from_end =
+        ((char*)end_point - 1) - (char*)p;
+      size_t end_to_end = round_by_sizeclass(rsize, (size_t)offset_from_end);
+      return address_cast<char>((char*)end_point_correction - end_to_end);
```

Comment:
Lots of C-style casts added here.  Please can you turn them back into teh correct kinds of C++ casts?